File: CPP/src/Formula.cpp
#ifndef ElaroSolutions_DARFormula_FORMULA_CPP
#define ElaroSolutions_DARFormula_FORMULA_CPP


#include "Formula.h"

namespace ElaroSolutions { namespace DARFormula {


    Formula::Formula()
    {
        _data = nullptr;
        _root = nullptr;
        _throwsExceptionFromCalcValue = false;
        addVariable("E",2.718281828459045);
        addVariable("PI",3.141592653589793);
        addVariable("PHI",1.618033988749895);
    }

    Formula *Formula::setUpFormula(std::initializer_list<std::string> allowedVariables,
                                   std::initializer_list<std::string> allowedFields, IDataStructure *data) {

        for(const std::string &variable : allowedVariables)
        {
            addVariable(variable,0.0);
        }

        for(const std::string &field: allowedFields)
        {
            addField(field);
        }

        setData(data);
        return this;
    }

    void Formula::checkVariablesAndFields()
    {
            checkNode(_root);
    }

    void Formula::checkNode(Node *n)
    {
        switch(n->getType())
        {
            case Simple:
            {
                if (typeid(*n) == typeid(VariableNode))
                {
                    if(_allowedVariables.find(n->toText())!=_allowedVariables.end())
                    {
                        return;
                    }
                    throw UnexpectedVariable("Unsupported variable : " + (n->toText()));

                }
                else if (typeid(*n) == typeid(DataNode))
                {
                    auto *d = (DataNode *)n;
                    for (auto index:d->getIndexes())
                    {
                        checkNode(index);
                    }
                    if((_allowedFields.empty() && d->getField().empty())||_allowedFields.find(d->getField())!=_allowedFields.end())
                    {
                        return;
                    }
                    throw UnexpectedVariable("Unsupported field : " + d->getField());
                }
            }
            break;
            case Unary:
            {
                checkNode(((UnaryNode *)n)->getOperand());
            }
            break;
            case Binary:
            {
                checkNode(((BinaryNode*)n)->getPreOperand());
                checkNode(((BinaryNode*)n)->getPostOperand());
            }
            break;
            case Ternary:
            {
                checkNode(((TernaryNode*)n)->getLimit());
                if(_allowedVariables.find(((TernaryNode*)n)->getCounter()->toText())!=_allowedVariables.end())
                {
                    throw UnexpectedVariable("Counting variable "+((TernaryNode*)n)->getCounter()->toText()+ " already in use");
                }
                else
                {
                    _allowedVariables.insert(((TernaryNode*)n)->getCounter()->toText());
                }
                checkNode(((TernaryNode*)n)->getFormula());
                _allowedVariables.erase(((TernaryNode*)n)->getCounter()->toText());
            }
            break;
            default:return;
        }

    }

    Formula* Formula::setFormula(const std::string& formula)
    {
        delete _root;

        _root = Parser::Parse(Parser::Tokenize(formula),&_variables);

        checkVariablesAndFields();
        giveDataToDataNodes(_root);
        return this;
    }

    void Formula::addVariable(const std::string& variableName, double initialValue)
    {
        if(_variables.count(variableName)>0)
        {
            updateVariable(variableName,initialValue);
            return;
        }
        _variables.emplace(variableName, initialValue);
        _allowedVariables.emplace(variableName);
    }

    double Formula::getVariableValue(const std::string& variableName)
    {
        return _variables.at(variableName);
    }

    Formula* Formula::updateVariable(const std::string& variableName, double value)
    {
        _variables.at(variableName)=value;
        return this;
    }

    Formula* Formula::addOneToVariable(const std::string& variableName)
    {
        _variables.at(variableName)++;
        return this;
    }

    void Formula::setData(IDataStructure *data) {
        _data =data;
    }

    void Formula::addField(const std::string& fieldName) {
        _allowedFields.insert(fieldName);
    }

    double Formula::calculateValue() {
        double result=0.0;
        try{
            result = _root->calcValue();
        } catch (std::exception &r)
        {
            if (!_throwsExceptionFromCalcValue)
                return NAN;
            else
             throw;
        }
        return result;
    }

    Formula* Formula::enableExceptionsOnCalculateValue() {
        _throwsExceptionFromCalcValue = true;
        return this;
    }

    Formula* Formula::disableExceptionsOnCalculateValue() {
        _throwsExceptionFromCalcValue = false;
        return this;
    }

    bool Formula::exceptionsOnCalculateValueEnabled() const {
        return _throwsExceptionFromCalcValue;
    }

    std::string Formula::formulaToText() {
        return _root->toText();
    }

    Formula::~Formula()
    {
        delete _root;

    }

    Formula* Formula::limitDataIndexQuantity(int quantity)
    {
        limitDataNode(quantity,_root);
        return this;
    }

    void Formula::limitDataNode(int quantity, Node *n)
    {
        if(n->getType()==Simple)
        {
            if(typeid(*n) == typeid(DataNode))
            {
                auto d = (DataNode *)n;
                if(d->getIndexes().size()<quantity)
                    throw BadFormula("Not enough bracketed formulas after data declaration");
                else
                    d->getIndexes().resize(quantity);
            }
            else return;
        }
        else if(n->getType()==Unary)
        {
            limitDataNode(quantity,((UnaryNode *)n)->getOperand());
        }
        else if(n->getType()==Binary)
        {
            limitDataNode(quantity,((BinaryNode *)n)->getPreOperand());
            limitDataNode(quantity,((BinaryNode *)n)->getPostOperand());
        }
        else if(n->getType()==Ternary)
        {
            limitDataNode(quantity,((TernaryNode *)n)->getLimit());
            limitDataNode(quantity,((TernaryNode *)n)->getFormula());
        }
    }

        void Formula::giveDataToDataNodes(Node *n)
        {
            if(typeid(*n) == typeid(DataNode))
            {
                auto d = (DataNode *)n;
                d->setData(_data);
                for(auto indexNode : d->getIndexes())
                {
                    giveDataToDataNodes(indexNode);
                }
            }
            else if(n->getType()==Unary)
            {
                giveDataToDataNodes(((UnaryNode *)n)->getOperand());
            }
            else if(n->getType()==Binary)
            {
                giveDataToDataNodes(((BinaryNode *)n)->getPreOperand());
                giveDataToDataNodes(((BinaryNode *)n)->getPostOperand());
            }
            else if(n->getType()==Ternary)
            {
                giveDataToDataNodes(((TernaryNode *)n)->getLimit());
                giveDataToDataNodes(((TernaryNode *)n)->getFormula());
            }
        }




    } }

#endif

File: CSharp/FormulaParser/Formula.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;

using ElaroSolutions.FormulaParser;


namespace ElaroSolutions.FormulaParser
{
    public class Formula
    {
        protected Dictionary<string, double> Variables;
        private readonly HashSet<string> _allowedVariables;
        private readonly HashSet<string> _allowedFields;
        private IDataLists _data;
        private Parser _parser;
        private Scanner _scanner;
        private Node _root;
        public string SyntaxError { get; private set; }
        public Boolean ThrowsExceptionsFromCalcValue { get; set; }

        public Formula()
        {
            Variables = new Dictionary<string, double>();
            _allowedFields = new HashSet<string>();
            _allowedVariables = new HashSet<string>() { "r" };
            ThrowsExceptionsFromCalcValue = false;
            _root = null;
            AddVariable("E", Math.E);
            AddVariable("PI", Math.PI);
            AddVariable("PHI", (1.0 + Math.Sqrt(5.0)) / 2.0);
        }

        public void SetUpFormula(IEnumerable<string> allowedVariables, IEnumerable<string> allowedFields, IDataLists data)
        {
            foreach (var variable in allowedVariables)
            {
                AddVariable(variable, 0.0);
            }

            foreach (var field in allowedFields)
            {
                AddField(field);
            }

            SetData(data);
        }

        public void SetData(IDataLists data)
        {
            this._data = data;
        }

        public void AddVariable(string variableName, double initialValue)
        {
            _allowedVariables.Add(variableName);
            Variables.Add(variableName, initialValue); 
        }

        public void AddField(string fieldName)
        {
            _allowedFields.Add(fieldName);
        }

        public virtual void SetFormula(string formula)
        {
            _root = null;
            SyntaxError = "";
            byte[] formulaInBytes = Encoding.UTF8.GetBytes(formula);
            MemoryStream formulaStream = new(formulaInBytes);
            _scanner = new Scanner(formulaStream);
            _parser = new Parser(_scanner, ref Variables, _data);
            _root = _parser.Parse();
            if (_root != null)
            {
                try { this.CheckVariablesAndFields(); }
                catch (UnexpectedVariableException u)
                { SyntaxError = u.Message; }

            }
            else SyntaxError = _parser.errors.errorsString;
        }

        private void CheckNode(Node n)
        {
            if (n is VariableNode node1)
            {
                if (_allowedVariables.Any(variable => String.Compare(node1.ToString(), variable, StringComparison.Ordinal) == 0))
                {
                    return;
                }
                throw new UnexpectedVariableException("Unrecognized variable: " + n.ToString());
            }
            else if (n is DataNode node)
            {
                foreach (Node n1 in node.DataIndexes)
                {
                    CheckNode(n1);
                }

                if (_allowedFields.Any(field =>
                    String.Compare(node.FieldToString(), field, StringComparison.Ordinal) == 0))
                {
                    return;
                }

                throw new UnexpectedVariableException("Unsupported field: " + node.FieldToString());
            }
            else if (n is UnaryNode unaryNode)
            {
                CheckNode(unaryNode.operand);
                return;
            }
            else if (n is BinaryNode binaryNode)
            {
                CheckNode(binaryNode.preoperand);
                CheckNode(binaryNode.postoperand);
                return;
            }
            else if (n is TernaryNode ternaryNode)
            {
                if (_allowedVariables.Contains(ternaryNode.counter.ToString()))
                {
                    throw new UnexpectedVariableException("Unusable counting variable: " + ternaryNode.counter.ToString() + " already in use");
                }
                else
                {
                    CheckNode(ternaryNode.limit);
                    _allowedVariables.Add(ternaryNode.counter.ToString());
                    CheckNode(ternaryNode.formula);
                    _allowedVariables.Remove(ternaryNode.counter.ToString());
                    return;
                }

            }

        }

        private void CheckVariablesAndFields()
        {
            CheckNode(_root);
        }

        public void UpdateVariable(string varName, double value)
        {
            Variables[varName] = value;
        }

        public double CalculateValue()
        {
            double result;
            if (SyntaxError.Length > 0)
            { throw new InvalidOperationException(); }
            try
            {
                result = _root.CalcValue();
            }
            catch (Exception)
            {
                if (ThrowsExceptionsFromCalcValue)
                {
                    throw;
                }
                else
                {
                    return double.NaN;
                }

            }
            if (Double.IsFinite(result))
                return result;
            else return double.NaN;
        }

        public string FormulaToString()
        {
            return _root.Equals(null) ? "" : _root.ToString();
        }
    }



}